### ISSUE TITLE
Documentation refinement... The Azure Service Principal Adventure

### ISSUE THREAD

**Issue by MiYanni** [Azure/azure-powershell#5182]
### User Scenario
I want to create an Azure Service Principal for running PowerShell scripts on Jenkins.

### The Story
So, the adventure starts here:
https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authenticate-service-principal
Scanning this document, it looks like it shouldn't take long... 15 mins tops. Looking at the three options, password, self-signed certificate, or certificate from Certificate Authority, I know the simplest path is to use password, because Jenkins has a section in Build Environment for storing passwords as environment variables. So, I dig into this section of the document. This is where my first snag occurs.

There are 2 scripts under the 'Create service principal with password' section. There is a short script and a long script. The short script seems easy and straightforward. The long script has a bunch of bullet points under it explaining various conditions and situations when using it. Anyone's natural inclination is to try to use the short script, since the long script sounds way more heavy handed than I need.

There is a statement above the long script:

> The following script enables you to specify a scope other than the **default subscription**, and retries the role assignment if an error occurs:

I've put the part in bold that was a red flag to me. If the long script works on a scope other than the default subscription, that probably means the short script only works on the default subscription (otherwise, why would there be a long script?). So, in this story, I am aware that I have multiple subscriptions. My questions now become, "What is a default subscription? How do I determine what is my default subscription?" etc., etc. Let's go further down the rabbit hole...

So, if you Google 'Azure Default Subscription', the first result is this:
https://docs.microsoft.com/en-us/powershell/module/azure/select-azuresubscription?view=azuresmps-4.0.0
Now, I have a bit of knowledge that commands using -Azure are RDFE, and ones using -AzureRM are ARM. However, I didn't know if this would affect subscriptions. So, I decide to try this command to set the default. It doesn't work. It shows an error to the effect of "the subscription name doesn't exist".

Next, I search for "the subscription name doesn't exist" and find this:
https://stackoverflow.com/questions/34961993/azure-powershell-select-azuresubscription-error-the-subscription-name-doesnt
Now, the answer to this states how to select a subscription, but I have no idea if that is actually setting the **default** subscription. It does, however, make me realize I need to find an AzureRM (ARM) way to set this.

My next search gets me here:
https://stackoverflow.com/questions/40978484/how-to-change-the-default-azure-rm-subscription
This question does not have a marked answer. But, I scroll through reading them anyway. I look at this one and read the steps:
https://stackoverflow.com/a/44131117/294804
I noticed Step 3:

> Step 3: (Get-AzureRmContext).Subscription
> Confirm that you have selected the right subscription.

This gives me one of those 'ah ha!' moments. I realize that this **default** subscription probably has to do with AzureRmContext. Now, until this point, I haven't had to change my context. I didn't know what it was. Then, I look at the first response on the StackOverflow question again:
https://stackoverflow.com/a/47675981/294804
It mentioned **default**. I put 2&2 together and realized, "Oh, the context is what my default is for the PowerShell session." Up until then, I thought **default subscription** was some kind of property in Azure itself for my account. Earlier, I had even looked at my subscriptions in the portal to see if I could find one that was _marked_ as **default**.

With that knowledge in-hand, I set my context to be the correct subscription, and proceed with creating the service principal and setting the role assignment. Unfortunately, my story doesn't end there. I scroll down to the section "Provide credentials through PowerShell". This will allow me to log in as that service principal.

I read this information prior to the 2-line script:

> Now, you need to log in as the application to perform operations. For the **user name**, use the ApplicationId that you created for the application. For the **password**, use the one you specified when creating the account.

Then I stare at the script:
`$creds = Get-Credential`
`Login-AzureRmAccount -Credential $creds -ServicePrincipal -TenantId {tenant-ID}`

I have bolded the text that was confusing to me. User name and password? Where are these in the script? How do I use them? I proceed to run the script. Low and behold, Get-Credential asks for a Username and Password. Dun dun daa. This prompts the question, "Well, how do I provide them programmatically?" since the Jenkins job can't have an interactive prompt.

My next Google search gets me here:
https://stackoverflow.com/questions/6239647/using-powershell-credentials-without-being-prompted-for-a-password
Looking at the answer, the first sentence states:

> The problem with Get-Credential is that it will always prompt for a password.

My response is, "What. What??? Why?" I didn't want to have to provide a file since I already have a way to supply a password directly via an environment variable (in Jenkins). I continue to search and look through the rest of the "Use Azure PowerShell to create a service principal to access resources". At the bottom, there is a section "Save access token to simplify log in". I hoped this would aid me. However, again, this saves a token of an AzureRmProfile (I still don't know what that is). I create one and looked at the json and concluded it would not be useful for my current situation.

I decide to look up Get-Credential directly, and arrive here:
https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.security/get-credential?view=powershell-5.1
I scroll through this page, trying to find anything of use. I find this section, "Example 7: Create a credential without prompting the user", and retort, "Omg, finally! My answer!" I read through the 3 lines 
 of script and exclaim, "Oh, you can just create the f*ing credential. Why didn't it tell me that earlier?!!"

With all that knowledge in-hand, I quickly look up how to get environment variables via PowerShell (very easy using $env:). So, I can create my simple login script and put it into Jenkins (Note: I've omitted specific, hard-coded data using %%):
`$password = ConvertTo-SecureString -String $env:%VAR% -AsPlainText -Force`
`$creds = New-Object -TypeName 'System.Management.Automation.PSCredential' -ArgumentList %APPID%, $password`
`Login-AzureRmAccount -Credential $creds -ServicePrincipal -TenantId %TENANTID%`

**PS**: Additionally, through my search, I arrived here:
https://docs.microsoft.com/en-us/powershell/azure/authenticate-azureps?view=azurermps-5.1.1
The section "Log in with a service principal" shows a script using Login-AzureRmAccount. I still have some confusion from this... because there is an additional parameter in it compared to the page I was following. That param is -ApplicationId. I'm still left with the question... "If my creds are generated using my ApplicationId as my username and I'm providing my creds using -Credential, why do I need to provide -ApplicationId?" The other example doesn't use that param and it doesn't seem necessary or provide additional info to the command.

### Overall
Something that should have taken me 15 minutes took me over 2+ hours because the documentation send me on a wild goose chase. References to RDFE terms (**default subscription**), lack of clarity in the examples in the process, and nearly hidden information (creating a credential by-hand in the 7th example of 9 on the Get-Credential webpage) made this task much longer than it needed to be.

**Comment by markcowl** [comment]
@MiYanni Please provide recommendations on specific documentation changes to remedy this.  For example, 
- which documents referred to 'default subscription'
- did either of the scripts work out of the box, or did you need to edit them to get them to work (and what documentation changes would be needed to fix them).
- what changes to cmdlets or cmdlet help would have fixed this?

**Comment by MiYanni** [comment]
@markcowl 
- Which documents referred to 'default subscription'?
  - https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authenticate-service-principal#create-service-principal-with-password
  - https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authenticate-service-principal#create-service-principal-with-self-signed-certificate
- Did either of the scripts work out of the box, or did you need to edit them to get them to work (and what documentation changes would be needed to fix them)?
  - Most example scripts work out-of-the-box, if the documentation around them is understood properly.
  - https://docs.microsoft.com/en-us/powershell/azure/authenticate-azureps?view=azurermps-5.1.1#log-in-with-a-service-principal
    - This script has a variable, `$pscredential`, which isn't described/assigned. So, the script will not work out-of-the-box. Would need to state how to assign this value.
  - In general, the interaction between logging in, credentials, subscriptions, and the context (when dealing with service principals) was a lacking. Even just some 'Note:' sentences with links to other documents about these topics would have been helpful. Additionally, making the documentation around Service Providers 'CI friendly' would have helped immensely.
  - https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authenticate-service-principal#create-service-principal-with-password
    - **Provide credentials through PowerShell** section shows how to create credentials (`Get-Credential`) when logging in interactively as the Service Principal. I'd assume that most people would not want interactive credential creation as they are probably using a Service Principal as part of a CI process.
      - See the **Example 7** mentioned below on how to handle non-interactive credential creation.
- What changes to cmdlets or cmdlet help would have fixed this?
  - _Related:_ https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.security/get-credential?view=powershell-5.1#examples
    - **Example 7: Create a credential without prompting the user** is a good example to show how to login using a credential, and without using an access token. 
      - I'd recommend showing this as part of this login process: https://docs.microsoft.com/en-us/powershell/azure/authenticate-azureps?view=azurermps-5.1.1#log-in-with-a-service-principal, or as part of this (need to make this section more generic for any non-interactive login via service principal): https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authenticate-service-principal#save-access-token-to-simplify-log-in 
      - From a new user's perspective, why can't `-AccountId` **and** `-Password` simply be parameters for `Login-AzureRmAccount`? That was the most confusing part for me. If I can generate a credential with those 3 lines of script, why can't `Login-AzureRmAccount` do that for me?
  - _Related:_ https://www.codeisahighway.com/how-to-easily-and-silently-obtain-accesstoken-bearer-from-an-existing-azure-powershell-session/
    -  This is also a very useful method for getting an access token based on the currently logged in account. I'd recommend we either make a cmdlet for this, or have it returned as part of the context.

**Comment by twitchax** [comment]
@sdwheeler for FYI.

**Comment by sdwheeler** [comment]
@tfitzmac for FYI - Tom owns the Azure doc [referenced](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authenticate-service-principal). 

I can look at updating the AzurePS [doc](https://github.com/Azure/azure-docs-powershell/blob/master/docs-conceptual/azurermps-5.1.1/authenticate-azureps.md).

**Comment by tfitzmac** [comment]
@MiYanni - thanks for this feedback. I am reviewing it, and looking at how to resolve it. One thing I think we need to do is reduce the number of articles that show similar content. That way we can provide simple, consistent examples.

**Comment by MiYanni** [comment]
@MiYanni Clean up the doc

**Comment by tfitzmac** [comment]
@MiYanni - I made a bunch of changes to [this doc](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authenticate-service-principal) that will get published this afternoon. To avoid duplication, I removed the section about authenticating with a password. For password, it now directs users to the [article](https://docs.microsoft.com/en-us/powershell/azure/create-azure-service-principal-azureps?view=azurermps-5.4.0) that is part of the PowerShell getting started docs.

I added more explanation about Set-AzureRmContext and selecting a subscription. 

I added a note in the intro that directs users to Azure AD MSI for supported scenarios. MSI is going to be a much better option for many users. 

My understanding has been that we shouldn't show code with an embedded password. That is why the login command does not have a password parameter. So, I would be hesitant about adding that to our examples.
